### ISSUE TITLE
RFC: Learning from data — how Talon rules adapt to each customer's reality

### ISSUE THREAD

**Issue by OpakAlex** [opentalon/talon-language#3]
## The Problem

Classical expert systems (and Prolog) have a fundamental weakness: **rules are static**. Someone writes "service interval = 20,000 km" and that number is wrong for half the customers. One fleet services at 15,000, another at 30,000. Hand-writing thresholds doesn't scale across thousands of tenants with different equipment, usage patterns, and maintenance cultures.

At the same time, pure ML approaches (neural nets, classifiers) lose what makes expert systems valuable: rules are **readable, explainable, auditable, and overridable**. A fleet manager can't inspect a neural net. They can inspect a Talon rule.

## The Goal

**Rules that write themselves from each customer's own data, but stay readable, explainable, and overridable.**

Three concrete mechanisms, all expressible as Talon rules:

---

## 1. Threshold Mining — Rules Stay, Numbers Learn

The rule *structure* is written by a human. The *thresholds* come from the customer's own historical data.

**Human writes the shape:**
\`\`\`talon
detect "Service overdue" {
  for records where category in category_tree("Vehicles")
    and attr "km" > attr "last_service_km" + THRESHOLD
  flag matching items
  priority HIGH
}
\`\`\`

**Talon learns THRESHOLD per customer from their ticket history:**
- Customer A: last 50 service tickets averaged 18,200 km intervals → THRESHOLD = 18,200
- Customer B: last 50 service tickets averaged 25,000 km → THRESHOLD = 25,000

The learning query is itself a Talon rule:

\`\`\`talon
define "learned_service_interval" {
  calculate avg_interval from tickets
    where attr "ticket_template_id" in service_templates
    group by item
    compute attr "km_at_service" - attr "previous_km_at_service"
    over last 12 months
  return avg_interval * 0.9   // 10% safety margin
}
\`\`\`

**Why this matters for Timly:** every customer has different fleet sizes, usage patterns, and maintenance cultures. A fixed 20,000 km is wrong for most. Their own ticket history is right.

### More examples of learnable thresholds:

| Rule | Fixed threshold (bad) | Learned threshold (good) | Learned from |
|------|----------------------|--------------------------|--------------|
| Service interval | 20,000 km | Customer's avg service interval | Ticket history |
| Stock reorder point | Fixed "minimum_amount" | Actual consumption rate * lead time | Consume activities |
| Ticket SLA | 7 days | Customer's avg resolution time * 1.2 | Closed ticket durations |
| "Idle asset" threshold | 90 days | Customer's avg assignment gap | Activity history |
| Consumption spike | 2x last week | Customer's own variance band | 90-day consumption stats |

---

## 2. Pattern Discovery — New Rules from Correlations

Beyond tuning thresholds, Talon can **discover entirely new rules** by mining correlations in historical data.

**Example: the engine scans 14 months of ticket + item + activity data and finds:**
> "Items in category 'Vans' that visit container 'External Workshop' have a 4x higher ticket rate within 60 days"

**Auto-generates a Talon rule:**
\`\`\`talon
detect "auto-discovered: External Workshop correlation" {
  for records where category == "Vans"
    and has activity where resource_name == "External Workshop"
    within last 30 days
  flag matching items
  label "{item.name} was at External Workshop recently — historically 4x failure rate"
  confidence 0.82
  source "mined from 14 months of data, 47 matching cases"
  priority MEDIUM
}
\`\`\`

**The key difference from ML:** the discovered rule is **readable**. A fleet manager can look at it and say "yes, that workshop does shoddy work" or "no, that's correlation not causation, delete it." They can edit, promote, or remove the rule.

### What pattern discovery looks for:

| Pattern type | What it finds | Example |
|-------------|---------------|---------|
| **Temporal correlation** | Event A followed by event B within N days | Electrical fault → engine failure within 30 days |
| **Attribute clustering** | Items with similar attributes fail together | Items with operating_hours > 2000 AND category "Pumps" |
| **Location correlation** | Items passing through location X fail more | External Workshop has 4x failure rate |
| **Seasonal patterns** | Consumption/failure spikes at certain times | Safety glove consumption doubles in winter |
| **Category anomaly** | One subcategory has unusual metrics vs siblings | Electric Vans fail 3x more than Diesel Vans |

---

## 3. Consumption Forecasting — Statistical Functions Inside Rules

Stock depletion isn't a fixed threshold — it's a trend. Talon needs built-in statistical primitives that **compute from data, not from hardcoded values:**

\`\`\`talon
detect "Stock empty in N days" {
  for records where type == "stock_item"
  calculate daily_rate from activities
    where type == "consume"
    weighted_moving_average last 30 days    // recent days weigh more
  calculate days_remaining = attr "current_stock" / daily_rate
  when days_remaining < 7
    and daily_rate > 0
  flag matching items
  label "{item.name}: ~{days_remaining} days left at current rate"
  priority HIGH
}
\`\`\`

### Statistical primitives needed in Talon:

| Function | Purpose | Example use |
|----------|---------|-------------|
| `avg` | Simple average | Average ticket resolution time |
| `weighted_moving_average` | Recent data weighs more | Consumption rate (Monday matters more than last month) |
| `stddev` | Variance detection | Is this week's consumption normal? |
| `trend` | Is it going up or down? | Stock level trending down faster than usual |
| `percentile` | Outlier detection | This item's failure rate is in the 95th percentile |
| `count_where` | Conditional counting | How many tickets in last 90 days |
| `correlation` | Two attributes move together? | km and failure rate correlation |

These are **not ML**. They're math. But they give rules the ability to **adapt to data** instead of relying on fixed constants.

---

## 4. The Feedback Loop — Rules That Learn from User Reactions

This is the piece classical Prolog never had. When Talon fires a detection, the user's reaction becomes training data:

\`\`\`
1. Talon fires: "VW Transporter overdue for service"

2. User reacts:
   a. Creates a ticket       → CONFIRMED  (strengthen rule)
   b. Dismisses the alert    → REJECTED   (weaken confidence)
   c. Ignores it             → UNKNOWN    (no signal)

3. Outcome later:
   a. Item actually broke    → RIGHT      (increase priority next time)
   b. Item ran fine 6 months → PREMATURE  (increase threshold)
\`\`\`

Feedback is stored as facts:
\`\`\`talon
// Feedback facts
feedback(entity, rule_id, "service_overdue", confirmed, "2025-05-06")
feedback(entity, rule_id, "service_overdue", rejected, "2025-04-15")

// Confidence computed from feedback
define "confidence_for" (rule_name) {
  calculate confirmed from feedback
    where rule == rule_name and result == "confirmed"
  calculate total from feedback
    where rule == rule_name
  return confirmed / total
}

// Weak rules auto-demote
detect "Weak auto-discovered rule" {
  for auto_discovered_rules
    where confidence_for(rule.name) < 0.3
    and total_feedback > 10
  demote to priority LOW
  label "Rule '{rule.name}' confirmed only {confidence}% — consider removing"
}
\`\`\`

**The feedback loop makes the system self-correcting.** Rules that fire correctly get promoted. Rules that annoy users get demoted. Thresholds drift toward each customer's reality.

---

## 5. The Three-Layer Architecture

All three mechanisms produce **Talon rules**. They look the same, run the same, can be inspected the same:

\`\`\`
Layer 1: STATIC RULES (human-written, 100% confidence)
├── "Defective item without ticket"         → always valid
├── "Certification expired"                 → compliance, non-negotiable
├── "Duplicate barcodes"                    → data integrity
└── Written by Timly team or MCP server authors

Layer 2: PARAMETERIZED RULES (human shape, learned thresholds)
├── "Service overdue by km"                 → shape fixed, km threshold learned
├── "Stock empty in N days"                 → shape fixed, consumption rate computed
├── "Ticket SLA breach"                     → shape fixed, SLA duration learned
└── Thresholds recalculated periodically (daily/weekly)

Layer 3: DISCOVERED RULES (mined from data, confidence-scored)
├── "External Workshop correlation"         → found by correlation mining
├── "Monday consumption spike"              → found by time-series analysis
├── "Electrical→Engine failure chain"       → found by sequence mining
└── Start at low confidence, promoted by feedback, pruned if never confirmed
\`\`\`

**The user doesn't know (or care) which layer a rule comes from.** They just see "your drill is probably going to break" — and they can ask **why**, because every rule is readable Talon, not a black-box model.

---

## Why This Solves Prolog's Historical Problem

| Classical Prolog weakness | How Talon addresses it |
|--------------------------|----------------------|
| Thresholds are hardcoded | Learned from each customer's data |
| Rules must be hand-written | Pattern discovery generates new rules |
| No adaptation over time | Feedback loop strengthens/weakens rules |
| Brittle to new domains | Statistical primitives handle trends and variance |
| No confidence scoring | Every detection carries confidence + source |
| Can't say "I don't know" | Low-confidence rules are demoted, not forced |

The insight: **keep the inference engine symbolic (readable, auditable, overridable) but let the rule parameters and even the rules themselves be generated from data.** Best of both worlds.

---

## Open Questions

1. **How often should thresholds recalculate?** Daily is probably right for stock; weekly for maintenance intervals. Per-rule configuration?
2. **What mining algorithms for pattern discovery?** Association rule mining (Apriori) for correlations, sequential pattern mining for temporal chains. Both are well-understood, non-ML algorithms.
3. **How to handle cold-start?** New customers have no history. Start with Layer 1 (static rules from MCP server) and sensible defaults. Layer 2+3 kick in after enough data accumulates (e.g., 30 days).
4. **Should discovered rules require human approval before firing?** Maybe: discovered rules start as "draft" (visible to admins only) and must be promoted to "active" before they surface to end users.
5. **Privacy:** threshold mining must be per-tenant. Customer A's failure patterns must never influence Customer B's rules. Per-entity isolation is structural in OpenTalon — same principle applies to Talon.
